File: node/common/metrics.js
import chalk from 'chalk'

const logFormat = (funcName, elapsedTime, message) => {
  const colors = {
    getOverdueJobs: chalk.bold.red,
    dispatchOverdue: chalk.bold.blue,
    scanGroup: chalk.bold.yellow,
    tableEmpty: chalk.bold.magenta
  }
  return `${colors[funcName](funcName)} `+ 
  chalk.bgGreen.bold(`${elapsedTime}ms`) +
  `: ${message}`
}

export default class Metrics{
  constructor(startTime, partitions){
    this.empty = false
    this.startTime = startTime
    this.partitionSet = new Set(partitions)
  }
  // ddb_service metric
  getOverdueJobs(elapsedTime, partition){
    const message = logFormat('getOverdueJobs', elapsedTime, `retrieve overdue schedules from partition ${partition}`)
    console.log(message)
  }

  // worker metric
  dispatchOverdue(elapsedTime, partition, count){
    if(!this.empty && count==0){
      this.tableEmpty(partition)
    }
    const message = logFormat('dispatchOverdue', elapsedTime, `fetch ${count} schedule(s) and dispatch to destination from partition ${partition}`)
    console.log(message)
  }

  scanGroup(elapsedTime, partitions, scanImmediately){
    const message = logFormat('scanGroup', elapsedTime, `iterate all ${partitions.length} partitions once. start next scan immediately = ${scanImmediately}`)
    console.log(message)
  }

  // 최초로 테이블이 빌 때까지의 시간 측정
  tableEmpty(partition){
    this.partitionSet.delete(partition)
    if (this.partitionSet.size==0){
      this.empty = true
      const message = logFormat('tableEmpty', Date.now()-this.startTime, 'dispatch done')
      console.log(message)
    }
  }
}

File: golang/metrics/metrics.go
package metrics

import (
	"fmt"
	"time"

	"github.com/fatih/color"
)

var (
	EMPTY          bool
	startTime      time.Time
	PARTITIONS     = []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	partitionEmpty = make(map[int]bool)
	stopwatch      = color.New(color.FgWhite, color.BgGreen).SprintFunc()
	red            = color.New(color.Bold, color.FgRed).SprintFunc()
	blue           = color.New(color.Bold, color.FgBlue).SprintFunc()
	yellow         = color.New(color.Bold, color.FgYellow).SprintFunc()
	magenta        = color.New(color.Bold, color.FgMagenta).SprintFunc()
)

func init() {
	for _, partition := range PARTITIONS {
		partitionEmpty[partition] = false
	}
	EMPTY = false
	startTime = time.Now()
	fmt.Printf("== %s ==", yellow(startTime))
}

func GetOverdueJobs(elapsedTime time.Duration, partition int) {
	fmt.Printf("[%s] %s : retreive overdue schedules from partition %d\n", red("GetOverdueJobs"), stopwatch(elapsedTime), partition)
}

func DispatchOverdue(elapsedTime time.Duration, partition int, count int) {
	if !EMPTY && count == 0 {
		tableEmpty(partition)
	}
	fmt.Printf("[%s] %s : fetch %d schedules(s) and dispatch to destination from partition %d\n", blue("DispatchOverdue"), stopwatch(elapsedTime), partition, count)
}

func ScanGroup(elapsedTime time.Duration, partitions []int, scanImmediately bool) {
	fmt.Printf("[%s] %s : iterate all %d partitions once. start next scan immediately = %t\n", magenta("ScanGroup"), stopwatch(elapsedTime), len(partitions), scanImmediately)
}

func tableEmpty(partition int) {
	partitionEmpty[partition] = true
	allEmpty := true
	for _, partition := range PARTITIONS {
		allEmpty = allEmpty && partitionEmpty[partition]
	}
	if allEmpty {
		EMPTY = true
		fmt.Printf("[%s] %s : table is empty\n", yellow("TableEmpty"), stopwatch(time.Since(startTime)))
	}
}
